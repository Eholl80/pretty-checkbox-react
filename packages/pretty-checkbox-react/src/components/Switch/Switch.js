// @flow

import * as React from 'react';
import classNames from 'classnames';
import { BaseInput } from '../BaseInput/BaseInput';

import { type PCRColors, type PCRColorsOutline, type SwitchFillModes } from '../../types';
import { mergeRefs, isChecked, isBoolean } from '../../utils/utils';
import { __handleCheckboxChange } from '../Checkbox/Checkbox';

export type SwitchState = boolean | string | any[];

export type SwitchProps = {
    /**
     * The type of the Switch.
     */
    type?: 'radio' | 'checkbox',

    /**
     * Specify one of the colors.
     */
    color?: PCRColors | PCRColorsOutline,

    /**
     * Specify one of the fill modes.
     */
    fill?: SwitchFillModes,

    /**
     * Controlled mode checkbox.
     */
    checked?: boolean,

    /**
     * Runs inside the default onChange handler for the Checkbox.
     *
     * Optionally return `false` from a synchronous action to prevent the
     * checkbox from being un/checked.
     */
    onChange?: (e: SyntheticMouseEvent<HTMLInputElement>) => ?boolean,

    /**
     * Set the checkbox as disabled.
     */
    disabled?: boolean,

    /**
     * Set the checkbox as locked. This will also set the
     * `aria-disabled` attribute to `true`.
     */
    locked?: boolean,

    /**
     * Pass additional class names to the `.pretty` div.
     */
    className?: string,

    /**
     * The `state` prop from` useSwitchState`.
     */
    state?: SwitchState,

    /**
     * The `setState` dispatcher from `useSwitchState`.
     */
    setState?: (state: SwitchState) => void,

    /**
     * Stub...
     */
    value?: string,

    /**
     * Additional HTML attributes to pass to the `input` element.
     * TODO: fix flow type after https://github.com/facebook/flow/pull/7569 ðŸ™„
     */
    htmlProps?: Object,

    /**
     * Text to show inside the label _or_ a render function
     * that allows you to render the internal pretty-checkbox
     * `.state` div.
     *
     * @example
     * <Switch>
     *     {({ color, icon }) => (
     *         <div className="state my-custom-class-name">
     *             {icon}
     *             <label>Something super special</label>
     *         </div>
     *     )}
     * </Switch>
     */
    children: React.Node | ((...props: Object) => React.Node),
};

export const useSwitchState = (initialState: SwitchState | (() => SwitchState) = false) => {
    const [state, setState] = React.useState<SwitchState>(initialState);

    return { state, setState };
};

export const Switch = React.forwardRef<SwitchProps, HTMLInputElement>(({
    type = 'checkbox',
    color,
    fill,
    checked: checkedProp,
    onChange,
    disabled,
    locked,
    children,
    className,
    value,
    state,
    setState,
    ...htmlProps
}, ref) => {
    const bool = isBoolean(value);
    const checked = isChecked(checkedProp, state, value);

    const switchRef = React.useRef(mergeRefs(ref, React.createRef()));

    const handleChange = React.useCallback(
        (e: SyntheticMouseEvent<HTMLInputElement>) => {
            if (disabled || locked) {
                e.preventDefault();
                return;
            }

            if (typeof onChange === 'function') {
                if (onChange(e) === false) {
                    return;
                }
            }

            if (typeof setState !== 'function') {
                return;
            }

            if (type === 'radio') {
                setState(bool ? !state : e.currentTarget.value);
            } else {
                __handleCheckboxChange({ setState, state, value, bool });
            }
        },
        [setState, state, disabled, locked, onChange]
    );

    return (
        <BaseInput
            type={type}
            role={type}
            locked={locked}
            className={classNames(
                'p-switch',
                {
                    [fill ? `p-${fill}` : '']: fill,
                },
                className
            )}
            disabled={disabled}
            aria-checked={!!checked}
            aria-disabled={disabled || locked}
            // $FlowFixMe
            onChange={handleChange}
            checked={checked || (state && state === value)}
            value={value}
            tabIndex="0"
            ref={switchRef}
            {...htmlProps}>
            {children}
        </BaseInput>
    );
});
