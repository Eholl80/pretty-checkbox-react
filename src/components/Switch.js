// @flow

import React from 'react';
import classNames from 'classnames';

import Input, { PREFIX } from './Input';
import type { InputProps } from './Input';

export type SwitchProps = {
    ...InputProps,

    /**
     * Specify animation for the `Switch` component.
     */
    animation?: "smooth" | "jelly" | "tada",

    /**
     * The shape of the `Switch` component.
     */
    shape?: "outline" | "fill" | "slim",

    /**
     * Provide a name when `type` is `radio`.
     */
    name?: string
};

function Switch(props: SwitchProps) {
    const { type, shape, className, name, inputProps, ...rest } = props;

    if (!shape && shape !== 'outline'  && shape !== 'fill' && shape !== 'slim') {
        throw new Error('Shape can be one of the following: outline, fill, or slim');
    }

    if (!type) {
        throw new Error('type is required to be set');
    }

    if (rest.animation && rest.animation !== 'smooth' && rest.animation !== 'jelly' && rest.animation !== 'tada') {
        throw new Error('Switch animations can be one of the following: smooth, jelly, or tada');
    }

    return (
        <Input className={classNames(`${PREFIX}switch`, className)}
            type={type}
            shape={shape}
            inputProps={{ ...inputProps, name: name }}
            {...rest} />
    );
}

Switch.defaultProps = {
    shape: 'outline',
    type: 'checkbox'
};

export default Switch;
