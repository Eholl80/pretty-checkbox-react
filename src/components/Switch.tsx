import * as React from 'react';

import { CommonProps, SwitchShape } from '../types/CommonProps';
import { createPrettyComponent } from '../factory/createPrettyComponent';
import { useCheckboxState, CheckboxState } from './Checkbox';
import { useRadioState, RadioState } from './Radio';

export type SwitchState = CheckboxState | RadioState;
export type SwitchProps = CommonProps<SwitchState> & {
    shape?: SwitchShape;
};

export const useSwitchState = ({
    type = 'checkbox',
    initialState = false,
}: {
    type?: CommonProps<SwitchState>['type'];
    initialState?: any;
} = {}) => {
    const checkbox = useCheckboxState(initialState);
    const radio = useRadioState(initialState);

    return Object.assign({ type }, type === 'checkbox' ? checkbox : radio);
};

export const Switch: React.FC<SwitchProps> = ({ value: userValue, ...rest }: SwitchProps) => {
    const value = typeof userValue === 'undefined' ? '' : userValue;

    return createPrettyComponent({
        type: 'checkbox',
        isSwitch: true,
        value,
        ...rest,
    });
};

Switch.displayName = 'Switch';
