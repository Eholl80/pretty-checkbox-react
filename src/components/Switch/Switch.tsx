import * as React from 'react';
import classNames from 'classnames';

import { BaseProps } from '../../typings/BaseProps';
import { BaseInput } from '../BaseInput/BaseInput';

import { useGenericChange, ChangeArgs } from '../../hooks/useGenericChange';
import { SwitchFillModes, SwitchAnimations } from '../../typings/types';
import { getChecked } from '../../utils/utils';
import { useGenericState } from '../../hooks/useGenericState';

export type SwitchState = boolean | string | any[] | undefined;

export type SwitchProps = BaseProps<SwitchState> & {
    /**
     * The name of the input field.
     */
    name?: string;

    /**
     * The type of Switch control. Can be a checkbox
     * or a radio.
     */
    type?: 'checkbox' | 'radio';

    /**
     * Specify one of the variants.
     */
    variant?: undefined;

    /**
     * Specify one of the fill modes.
     */
    fill?: SwitchFillModes;

    /**
     * Specify one of the animations.
     */
    animation?: SwitchAnimations;
};

export const useSwitchState = (initialState: SwitchState = false) =>
    useGenericState<SwitchState>(initialState);

const handleChange = ({ value, setState }: ChangeArgs<SwitchState>) => {
    if (typeof value === 'string') {
        setState(value);
    }
};

export const Switch = React.forwardRef(
    (
        {
            state,
            setState,
            value,
            disabled,
            locked,
            onChange,
            className,
            type = 'checkbox',
            ...props
        }: SwitchProps,
        htmlRef: React.Ref<HTMLInputElement>
    ) => {
        const defaultOnChange = useGenericChange<SwitchState>({
            setState,
            value,
            disabled,
            locked,
            onChange: handleChange,
        });

        const checked = React.useMemo(() => {
            if (typeof value === 'string') {
                return value === state;
            }

            return getChecked<SwitchState>({ value, state, setState });
        }, [value, state, setState]);

        return (
            <BaseInput
                type={type}
                state={checked}
                setState={setState}
                value={value}
                disabled={disabled}
                locked={locked}
                onChange={onChange || defaultOnChange}
                className={classNames('p-switch', className)}
                ref={htmlRef}
                {...props}
            />
        );
    }
);

Switch.displayName = 'Switch';
