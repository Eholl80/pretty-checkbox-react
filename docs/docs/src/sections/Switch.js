import React from 'react';

import { Switch as ReactSwitch } from 'pretty-checkbox-react';

import { Title } from '../components/Title';
import { CollapseContainer } from '../components/CollapseContainer';

import {
    LiveProvider,
    LiveEditor,
    LiveError,
    LivePreview
} from 'react-live';

const scope = { ReactSwitch };

const basicSwitch = `<>
    <ReactSwitch>Outline</ReactSwitch>
    <ReactSwitch shape="fill">Fill</ReactSwitch>
    <ReactSwitch shape="slim">Slim</ReactSwitch>
</>`;

function Switch() {
    return (
        <>
            <Title>Switch</Title>
            <div className="content">
                <p>Use the <strong>Switch</strong> component to start using switches as a checkbox or a radio.
                Customize the feel of switches via <strong>shape</strong> props. Use any of the following values:
                {' '}<strong>outline</strong> (default), <strong>fill</strong>, or <strong>slim</strong>.</p>
            </div>
            <LiveProvider code={basicSwitch} scope={scope}>
                <CollapseContainer title="iOS Style" demo={<LivePreview />}>
                    <>
                        <LiveEditor />
                        <LiveError />
                    </>
                </CollapseContainer>
            </LiveProvider>
        </>
    );
}

export default Switch;
