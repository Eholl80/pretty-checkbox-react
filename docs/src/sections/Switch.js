import React from 'react';

import { Checkbox } from 'pretty-checkbox-react';

import { Title } from '../components/Title';
import { CollapseContainer } from '../components/CollapseContainer';

import {
  LiveProvider,
  LiveEditor,
  LiveError,
  LivePreview
} from 'react-live';

const scope = { Checkbox };

const basicSwitch = `<>
    <Checkbox className="p-switch">Outline</Checkbox>
    <Checkbox className="p-switch p-fill">Fill</Checkbox>
    <Checkbox className="p-switch p-slim">Slim</Checkbox>
</>`;

function Switch() {
    return (
        <>
            <Title>Switch</Title>
            <div className="content">
                <p>Add class <strong>p-switch</strong>. For shapes add class, <strong>p-outline</strong> or <strong>p-fill</strong> or <strong>p-slim</strong></p>
            </div>
            <LiveProvider code={basicSwitch} scope={scope}>
                <CollapseContainer title="iOS Style" demo={<LivePreview />}>
                    <>
                        <LiveEditor />
                        <LiveError />
                    </>
                </CollapseContainer>
            </LiveProvider>
        </>
    );
}

export default Switch;