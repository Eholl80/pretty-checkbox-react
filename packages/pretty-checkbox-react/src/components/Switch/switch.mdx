---
name: Switch
---

import { Switch, useSwitchState } from './Switch';
import { Playground, Props } from 'docz';

# Switch

<Props of={Switch} />

## Intro
Switches are an interested pattern that was popularized by mobile operating systems like iOS and Android. Switches are typically associated with simple
boolean values (e.g. yes/no) and they offer a way to break out of the banalities of using checkboxes. PCR allows you to use Switch either as
a checkbox or as a radio; however, by default Switch is configured to act as a **checkbox**.

While using this can be exciting, there are a few things that need to considered:

* When used as a Checkbox
    * No `mixed` or `indeterminate` state
    * Not valid to represent a tree-like structure of items
* When used as a Radio/Radio Group
    * Works as a series of radio options, but will not scale to more than a few options
    * Can lead to a confusing experience

Finally, the examples below serve as a means to _highlight features_ of the Switch component. They do not serve as UX guidelines.

### Switch Differences
Switch has a few key differences from Checkbox and Radio worth noting:

* No `icon` prop
* `fill` modes are different (see example)
* `type` prop exposed, can be `checkbox` or `radio`

## Basic Switch with Hooks

<Playground>
    {() => {
        // because switch is a reserved word
        const switchHook = useSwitchState();

        return (
            <Switch {...switchHook}>
                Basic Switch with Hooks
            </Switch>
        );
    }}
</Playground>

## Basic Switch without Hooks
Similar to Radio and Checkbox, Switch can be used without hooks if you're still in the process of migrating to hooks.

## Switch as a Radio
Switch can be used as a checkbox or a radio. Just specify the `type` prop and everything it taken care of for you.

<Playground>
    {() => {
        // because switch is a reserved word
        const switchHook = useSwitchState();

        return (
            <Switch type="radio" {...switchHook}>
                Basic Switch as a Radio
            </Switch>
        );
    }}
</Playground>

## Fill Modes
Switch has different fill modes compared to Radio and Checkbox. The three choices are:

* `outline` (default)
* `fill`
* `slim`

<Playground>
    {() => {
        // because switch is a reserved word
        const switchHook = useSwitchState();

        return (
            <>
                <Switch type="radio" value="outline" {...switchHook}>
                    Outline
                </Switch>
                <Switch type="radio" fill="fill" value="fill" {...switchHook}>
                    Fill
                </Switch>
                <Switch type="radio" fill="slim" value="slim" {...switchHook}>
                    Slim
                </Switch>
            </>
        );
    }}
</Playground>


## Composing Switches using Hooks
One of the patterns mentioned above is using the Switch component in a settings page. Since we can use the switch as a checkbox _or_ radio
we can compose one or more hooks to represent the state of the entire page. Let's see an example:

<Playground>
    {() => {
        const switchState = useSwitchState(["enableEmail", "enablePush"]);

        return (
            <table
                style={{
                    borderCollapse: 'collapse',
                    width: '50%'
                }}>
                <tbody>
                    <tr style={{ borderBottom: '1px solid #ccc' }}>
                        <td>Enable email notifications</td>
                        <td>
                            <Switch value="enableEmail" {...switchState} />
                        </td>
                    </tr>
                    <tr style={{ borderBottom: '1px solid #ccc' }}>
                        <td>Enable push notifications</td>
                        <td>
                            <Switch value="enablePush"{...switchState} />
                        </td>
                    </tr>
                    <tr style={{ borderBottom: '1px solid #ccc' }}>
                        <td>Allow app to use location</td>
                        <td>
                            <Switch value="allowLoc" {...switchState} />
                        </td>
                    </tr>
                    <tr>
                        <td>Display avatar</td>
                        <td>
                            <Switch value="useAvatar" {...switchState} />
                        </td>
                    </tr>
                </tbody>
            </table>
        );
    }}
</Playground>
