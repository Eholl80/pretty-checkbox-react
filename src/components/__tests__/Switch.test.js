import React from 'react';

import { render, cleanup, getByLabelText, getByTestId, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Switch from '../Switch';

afterEach(cleanup);

describe('Switch tests', function () {
    describe('Switch usage as a checkbox (default behavior)', function () {
        it('should work with basic usage', function () {
            const handleChange = jest.fn();
            const { container } = render(<Switch id="demo" onChange={handleChange}>hello there</Switch>);

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-outline');
            expect(getByTestId(container, 'pcr-input')).toHaveAttribute('id', 'demo');
            expect(getByLabelText(container, 'hello there')).toBeDefined();

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);
        });
    });

    describe('Switch usage as a radio', function () {
        it('should behave as a radio', function () {
            const handleChange = jest.fn();
            const { container } = render(<Switch id="foo" shape="slim" type="radio" onChange={handleChange} />);

            expect(getByTestId(container, 'pcr-input')).toHaveAttribute('type', 'radio');

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);
            expect(container).toMatchSnapshot();
        });

        it('should behave as a radio group', function () {
            const handleChange = jest.fn();

            const { container } = render(
                <>
                    <Switch id="demo" type="radio" name="radio-1" value="r1" onChange={handleChange}>Hello there.</Switch>
                    <Switch id="demo" type="radio" name="radio-1" value="r2" onChange={handleChange}>General Kenobi. You are a bold one.</Switch>
                    <Switch id="demo" type="radio" name="radio-1" value="r3" onChange={handleChange}>Your move.</Switch>
                </>
            );

            const r1 = container.firstChild.firstChild;
            const r2 = container.firstChild.nextSibling.firstChild;
            const r3 = container.firstChild.nextSibling.nextSibling.firstChild;

            fireEvent.click(r1);

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(r1.checked).toBe(true);

            fireEvent.click(r2);

            expect(handleChange).toHaveBeenCalledTimes(2);
            expect(r1.checked).toBe(false);
            expect(r2.checked).toBe(true);

            fireEvent.click(r3);

            expect(handleChange).toHaveBeenCalledTimes(3);
            expect(r1.checked).toBe(false);
            expect(r2.checked).toBe(false);
            expect(r3.checked).toBe(true);
            expect(container).toMatchSnapshot();
        });
    });

    describe('Switch errors', function () {
        it('should throw when type is not defined', function () {
            expect(() => render(<Switch id="demo" type={null}>hello there</Switch>)).toThrow();
        });

        it('should throw when shape is undefined', function () {
            expect(() => render(<Switch id="demo" shape={null}>hello there</Switch>)).toThrow();
        });

        it('should throw when an incorrect shape is used', function () {
            expect(() => render(<Switch id="demo" shape="curve">hello there</Switch>)).toThrow();
        });

        it('should throw when an incorrect animation is used', function () {
            expect(() => render(<Switch id="demo" animation="pulse" >hello there</Switch>)).toThrow();
        });
    });

    describe('Switch ref tests', function() {
        it('should accept custom refs', function () {
            const ref = React.createRef();

            render(<Switch ref={ref} />);
            expect(ref.current).toBeDefined();
            expect(ref.current).toBeInstanceOf(HTMLInputElement);
        });
    });
});
