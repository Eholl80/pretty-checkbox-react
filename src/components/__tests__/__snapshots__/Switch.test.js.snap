// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch tests Switch usage as a radio should behave as a radio 1`] = `
<div>
  <div
    class="pretty p-switch p-slim"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="foo"
      type="radio"
      value=""
    />
    <div
      class="state"
      data-testid="pcr-state"
    />
  </div>
</div>
`;

exports[`Switch tests Switch usage as a radio should behave as a radio group 1`] = `
<div>
  <div
    class="pretty p-switch p-outline"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r1"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        Hello there.
      </label>
    </div>
  </div>
  <div
    class="pretty p-switch p-outline"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r2"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        General Kenobi. You are a bold one.
      </label>
    </div>
  </div>
  <div
    class="pretty p-switch p-outline"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r3"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        Your move.
      </label>
    </div>
  </div>
</div>
`;
